### PR TITLE
BUG: infinite recursion when using pickle or copy

### PR DESCRIPTION
**Pull request by bergtholdt**
Fixes issue #1681

### CODE REVIEW

**Comment by jakevdp** [comment]
Looks good – can you add some tests of the behavior this is intended to fix, in order to prevent future regressions?

**Comment by bergtholdt** [comment]
Done. 

I found another instance in the tools directory. Why is there file-duplication in the source tree under /tools and altair/utils?

**Comment by jakevdp** [comment]
Take a look at the note at the top of the files in ``altair/utils`` - they are generated from the source in ``tools`` when you run ``tools/generate_schema_wrapper.py`` to generate the core of the altair codebase.

**Comment by bergtholdt** [comment]
OK. no changes for schemapi. Just some other stuff (docstrings in core.py). Not related so I won't commit anything. Please merge!

File: altair/utils/tests/test_schemapi.py [hunk 1]
```diff
@@ -212,6 +212,7 @@ def test_undefined_singleton():
 
 
 def test_copy():
+    from copy import copy as cpy, deepcopy
```
Comment (jakevdp) [inline altair/utils/tests/test_schemapi.py]: Please put imports at the top of the module.

File: altair/utils/tests/test_schemapi.py [hunk 2]
```diff
@@ -245,6 +246,31 @@ def test_copy():
     assert mydct['b'] == dct['b']
     assert mydct['c'] == dct['c']
 
+    # Using copy module
```
Comment (jakevdp) [inline altair/utils/tests/test_schemapi.py]: Please make this a separate test, as it is testing a different copy mechanism (it helps to isolate potential test failures).

**Comment by jakevdp** [comment]
I'd like to get this fix in soon for the 3.3 release – will you be able to work on it in the next couple days? If not I can take over.

**Comment by jakevdp** [comment]
Merged in #1805. Thanks!

**Pull request closed without merging**